### PR TITLE
CLI Argument Addition

### PR DESCRIPTION
**Pull request by codenamewei**
# How to run 
1. Build jar on terminal / command prompt
(Linux)
`./startClassifai.sh `
(Windows)
`startClassifai.bat
`

# Changes

1. [CLI Argument simplify and add on for creation of new project when starting project.
](https://github.com/CertifaiAI/classifai/blob/cw_v2_cli_refactoring/classifai-api/src/main/java/ai/classifai/config/CLIArgument.java)

**How to use the cli**
go to where classifai-uberjar-dev.jar is created 
Normally at _classifai-uberjar/target/_

**Run**
**unlockdb**
`java -jar classifai.jar --unlockdb`

**Create empty project with project name when starting classifai** (no data path)
`java -jar classifai.jar --newproject=helloworld --projecttype=segmentation`

**Create empty project with randomly assigned project name when starting classifai**
`java -jar classifai.jar --projecttype=segmentation`

**Create a project with data path** 
`java -jar classifai.jar --projecttype=segmentation --datapath=/Users/john/Desktop/sample-image`

2. license header 2020-2021 added

3. [docker file for sonar build improve](https://github.com/CertifaiAI/classifai/blob/b4271712a26a73eb3e0c537054621c644b4d1a0b/.github/workflows/build.yml)

4. docker mode (still work in progress, for now only cli part is done)
for now, if run with --docker, Welcome Launcher and Selector for files and folder will not show up. 
This is due to in docker environment, its not possible to select data point from gui. 
Welcome Launcher is not supported due to awt packages. Initiating it will result in error. 

@devenyantis @YCCertifai 
Appreciate if you guys can run through it, highlight if any issue, 
else the last one who approve the PR merge it. Thanks.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Tested on?

- [x] Windows  
- [ ] Linux Ubuntu 
- [ ] Centos 
- [x] Mac  
- [ ] Others  (State here -> xxx )  

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged

### CODE REVIEW

**Comment by YCCertifai** [comment]
everything is fine in for new CLI feature..
Just --newproject should be --projectname
**tested on linux

**Comment by YCCertifai** [comment]
another minor issue is I can create multiple project (same annotation type) with same name in CLI
But not able to do so in the GUI
Good about it is it does not overwrite previous data (with the same name).

**Comment by YCCertifai** [comment]
Screenshot from 2021-01-21 14-43-27

**Comment by codenamewei** [comment]
hmmm by right same name should appear just once...ill check about that 
thanks for highlighting @YCCertifai

File: classifai-core/src/main/java/ai/classifai/database/portfolio/PortfolioVerticle.java [hunk 1]
```diff
@@ -356,40 +373,56 @@ public void configurePortfolioVerticle()
...
+                    loadProjectLoader(projectIDList);
 
-                        portfolioDbClient.queryWithParams(PortfolioDbQuery.loadDbProject(), projectIDJson, projectNameFetch -> {
+                }
 
-                            if (projectNameFetch.succeeded()) {
-                                ResultSet resultSet = projectNameFetch.result();
+                ////create new project from cli if any after load from memory
+                configProjectFromCLI();
+            }
```
Comment (YCCertifai) [inline classifai-core/src/main/java/ai/classifai/database/portfolio/PortfolioVerticle.java:383]: I tried debug for a while, find that at this point, BB verticle and Seg verticle have not deployed yet(  means havent connected to database yet)
So the dictionary in ProjectHandler is still empty.
Maybe...
We can try to create the project after all vertices have been deployed

Comment (codenamewei) [reply classifai-core/src/main/java/ai/classifai/database/portfolio/PortfolioVerticle.java:383]: complications from asynchronous operations...

the logic of the code was put to first load existing projects then create https://github.com/CertifaiAI/classifai/blob/cw_v2_cli_refactoring/classifai-core/src/main/java/ai/classifai/database/portfolio/PortfolioVerticle.java#L377-L382

but unfortunately when creating project..the streams of information from database were still not loaded to the ProjectLoader yet..

will push up the fix very soon...

**Comment by devenyantis** [comment]
I am able to run the first 3 commands.

Just want to ask for the last one
`java -jar classifai.jar --projecttype=segmentation --datapath=/Users/john/Desktop/sample-image`
What should be the behavior for this? The image data are loaded directly when we open the project?

**Comment by codenamewei** [comment]
> I am able to run the first 3 commands.
> 
> Just want to ask for the last one
> `java -jar classifai.jar --projecttype=segmentation --datapath=/Users/john/Desktop/sample-image`
> What should be the behavior for this? The image data are loaded directly when we open the project?

yes , cause in docker the building jar and project opening need to happen once the docker is up

@devenyantis i've found a complication and still fixing it, 

ill ping you again to review the pr again next week

**Comment by codenamewei** [comment]
@devenyantis @YCCertifai im done with the modification 

```
`java -jar classifai.jar --projecttype=segmentation --datapath=/Users/john/Desktop/sample-image`
What should be the behavior for this? The image data are loaded directly when we open the project?
```

for the comment above. 
it would be two scenarioss 
1. if project not exist, generate a random project name in the realm of segmentation and load the data path
2. if project exist, load the project and add on the data path

**Review by YCCertifai — APPROVED** [review]
Everything works fine for me now
- project with same new cannot be generated multiple times.
- datapath arg works fine for new project and existing project.

**Comment by devenyantis** [comment]
First 3 are okay. Tested on Windows and Ubuntu 18.04 (VM)

However, for the` --datapath` argument...
- on Windows, I couldn't load the images
- on Ubuntu, the data load sometimes. like 2 out of 5 times

And there is one time I accidentally load a valid path but empty and it loads some images.
im1
This behavior doesn't happen all the time.

**Comment by codenamewei** [comment]
> First 3 are okay. Tested on Windows and Ubuntu 18.04 (VM)
> 
> However, for the` --datapath` argument...
> 
> * on Windows, I couldn't load the images
> * on Ubuntu, the data load sometimes. like 2 out of 5 times
> 
> And there is one time I accidentally load a valid path but empty and it loads some images.
> im1
> This behavior doesn't happen all the time.

ah..missed this message as no pings. will check into this tomorrow

**Comment by codenamewei** [comment]
> First 3 are okay. Tested on Windows and Ubuntu 18.04 (VM)
> 
> However, for the` --datapath` argument...
> 
> * on Windows, I couldn't load the images
> * on Ubuntu, the data load sometimes. like 2 out of 5 times
> 
> And there is one time I accidentally load a valid path but empty and it loads some images.
> im1
> This behavior doesn't happen all the time.

Hi @devenyantis , 

for windows, 
i have to do something like this, can you check if it works?

`java -jar classifai-uberjar-dev.jar --unlockdb --projecttype=boundingbox --datapath="C:\\Users\\chiaw\\Desktop\\sampleImages\\chip"`

for ubuntu, 
can you check again, i pushed some codes after the last time you tried..
if you faced it again, give me a more thorough walkthrough on how to reproduce the problem so i can solve it.

**Comment by devenyantis** [comment]
Hi @codenamewei 

I've tested the latest one.

- datapath arg works for both Windows and Ubuntu18 only if I choose --projecttype=boundingbox. 
- for --projecttype=segmentation, the program stuck at loading screen

Also, the problem where the program loads images given empty valid folder is gone.

**Comment by codenamewei** [comment]
@devenyantis thanks...ill recheck this.....

**Comment by codenamewei** [comment]
@devenyantis kindly check if the latest update solve the problem, thanks

**Comment by devenyantis** [comment]
@codenamewei 

- Stuck at loading screen is solved, but the data is not loaded for --projecttype=segmentation
- All good for --projecttype=boundingbox

**Comment by devenyantis** [comment]
@codenamewei 

I have pulled the latest version and rerun startClassifai.bat. But still not working for the segmentation arg.

`java -jar classifai-uberjar-dev.jar --unlockdb --projecttype=segmentation --datapath="C:\\Users\\Deven.Yantis\\Documents\\classifai-test-images"`

Images used
https://drive.google.com/drive/folders/1qV9EukLtEW-rCqiLpgL6yuNhPVLZSs17?usp=sharing

java version, output from `java -version`
`openjdk version "14.0.2" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.2+12)
OpenJDK Client VM AdoptOpenJDK (build 14.0.2+12, mixed mode, sharing)`

I also try delete .classifai so that the program rebuild new database but no luck.

For bounding box, only change --projecttype=boundingbox and it work.
`java -jar classifai-uberjar-dev.jar --unlockdb --projecttype=boundingbox --datapath="C:\\Users\\Deven.Yantis\\Documents\\classifai-test-images" `

**Comment by codenamewei** [comment]
@devenyantis mind to attach the classifai.log here?

in windows is in the path of C:\\Users\\Deven.Yantis\\.classifai\\logs\classifai.log

**Comment by devenyantis** [comment]
@codenamewei 

[classifai.log](https://github.com/CertifaiAI/classifai/files/5877607/classifai.log)

**Comment by codenamewei** [comment]
@YCCertifai need your help to see if you can reproduce deven issue 

i can see it can detect the folder image , and its working on my mac and windows, see if you can reproduce the issue 

```
Jan 27, 2021 10:44:44 am INFO  ai.classifai.config.CLIArgument - Data path imported through cli: C:\\Users\\Deven.Yantis\\Documents\\classifai-test-images
Jan 27, 2021 10:44:45 am INFO  ai.classifai.MainVerticle - Existing database of classifai on C:\Users\Deven.Yantis\.classifai
Jan 27, 2021 10:44:46 am INFO  a.c.d.portfolio.PortfolioVerticle - Create project (from cli) with name: amazing_spiderman in SEGMENTATION project.
```

**Comment by YCCertifai** [comment]
@codenamewei will try on linux and windows.

**Comment by YCCertifai** [comment]
@devenyantis @codenamewei  it's working in my windows and ubuntu too

**Comment by YCCertifai** [comment]
@codenamewei  I will help @devenyantis to find out the problem...

**Comment by codenamewei** [comment]
> @codenamewei I will help @devenyantis to find out the problem...

thanks , let me know if need my input

**Comment by devenyantis** [comment]
@codenamewei @YCCertifai 

I found out that [this line](https://github.com/CertifaiAI/classifai/blob/cw_v2_cli_refactoring/classifai-core/src/main/java/ai/classifai/util/data/ImageHandler.java#L252) cause the problem on my computer. Most of the time `SegVerticle.getJdbcClient()` is returning null which it suppose not to.

I've added few logs to help me see the process. Denoted by `DEVEN === some log ===`

**Failure case.** 

fail

The `CHECKING JDBC CLIENT` printed null.

**Success case**

success

The `CHECKING JDBC CLIENT` printed a value. The program works as expected when this happen.

I try some simple workaround to solve the problem.

```
JDBCClient jdbcValue = null;
while(jdbcValue == null) {
    log.info("== jdbcValue still null ===");
    jdbcValue = SegVerticle.getJdbcClient();
}
log.info("DEVEN === CHECKING JDBC CLIENT ===" + " " + jdbcValue);
SegVerticle.updateUUID(jdbcValue, SegDbQuery.createData(), projectID, filesCollection.get(i), uuid, i + 1);

```

The program works consistently this time.

The log. The jdbc client is null for sometimes.

workaround

Another success run log. Seems like the null result is random.

workaround2

Hope that helps.

**Comment by codenamewei** [comment]
sweet , i like how this is an issue we never seen before..

@YCCertifai do try to look into this.

after migration to h2, we should also see if this problem reoccurs....

alternatively, if it cant be solved, migration to h2 can be happens first and see if this problem still persist

**Comment by YCCertifai** [comment]
Ok.. will look into this and try to solve it...
If unable to solve on this week, then I'll proceed to database migration first.

**Review by YCCertifai — APPROVED** [review]

**Pull request merged**